### DISCUSSION TITLE
an APE Compiler

### DISCUSSION THREAD

**Discussion by ArakawaHenri** [jart/cosmopolitan#698] [Q&A]
While this sounds bold, I mean, is it possible to compile any compiler (could be modified specifically for cosmopolitan) as Actual Portable Executable? In this way, we will have a complete and truly cross-platform portable cosmopolitan ecosystem.

**Comment by mingodad** [comment]
Look at this issue https://github.com/jart/cosmopolitan/issues/61#issuecomment-1252238996 it somehow already discuss this.
I found right now that `pcc` (http://pcc.ludd.ltu.se/ ) can be a god candidate.

**Comment by ahgamut** [comment]
>  is it possible to compile any compiler (could be modified specifically for cosmopolitan) as Actual Portable Executable?

Yes! I patched `gcc-11` to simplify compiling software with Cosmopolitan Libc: https://ahgamut.github.io/2023/07/13/patching-gcc-cosmo/

My patch is here: https://github.com/ahgamut/gcc/tree/portcosmo-11.2

With this patch, all you need to do is provide the right flags to `./configure` or `cmake`, and use `objcopy` to get an Actually Portable Executable! See examples here: https://github.com/ahgamut/superconfigure

You can build the patched `gcc` with my fork of `musl-cross-make`: https://github.com/ahgamut/musl-cross-make/tree/portcosmo

If you want `gcc-11` _itself_ as an Actually Portable Executable, use `cosmocc` and `cosmoc++` from the Cosmopolitan repo, and use my patch here: https://github.com/ahgamut/musl-cross-make/tree/gccbuild

**Reply by ahgamut** [reply]
here's `gcc-11.2` as an Actually Portable Executable: https://github.com/ahgamut/musl-cross-make/releases/tag/z0.0.0
